python/semantic_kernel/agents/agent.py

```diff
@@ -478,3 +560,336 @@ def as_mcp_server(
+    """
+
+    def decorator(cls: type[_TAgent]) -> type[_TAgent]:
+        AGENT_TYPE_REGISTRY[agent_type.lower()] = cls
+        return cls
+
+    return decorator
+
+
+class AgentRegistry:
+    """Responsible for creating agents from YAML, dicts, or files."""
+
+    @staticmethod
+    def register_type(agent_type: str, agent_cls: type[Agent]) -> None:
+        """Register a new agent type at runtime.
+
+        Args:
+            agent_type: The string identifier representing the agent type (e.g., 'chat_completion_agent').
+            agent_cls: The class implementing the agent, inheriting from `Agent`.
+
+        Example:
+            AgentRegistry.register_type("my_custom_agent", MyCustomAgent)
+        """
+        AGENT_TYPE_REGISTRY[agent_type.lower()] = agent_cls
+
+    @staticmethod
+    async def create_from_yaml(
+        yaml_str: str,
+        *,
+        kernel: Kernel | None = None,
+        settings: "KernelBaseSettings | None" = None,
+        extras: dict[str, Any] | None = None,
+        **kwargs,
+    ) -> _TAgent:
+        """Create a single agent instance from a YAML string.
+
+        Args:
+            yaml_str: The YAML string defining the agent.
+            kernel: The Kernel instance to use for tool resolution and agent initialization.
+            settings: The settings to use for the agent.
+            extras: Additional parameters to resolve placeholders in the YAML.
+            **kwargs: Additional parameters passed to the agent constructor if required.
+
+        Returns:
+            An instance of the requested agent.
+
+        Raises:
+            AgentInitializationException: If the YAML is invalid or the agent type is not supported.
+
+        Example:
+            agent = await AgentRegistry.create_agent_from_yaml(
+                yaml_str, kernel=kernel, service=AzureChatCompletion(),
+            )
+        """
+        data = yaml.safe_load(yaml_str)
+
+        agent_type = data.get("type", "").lower()
+        if not agent_type:
+            raise AgentInitializationException("Missing 'type' field in agent definition.")
+
+        if agent_type not in AGENT_TYPE_REGISTRY:
+            raise AgentInitializationException(f"Agent type '{agent_type}' not registered.")
+
+        if not kernel:
+            kernel = Kernel()
+
+        agent_cls = AGENT_TYPE_REGISTRY[agent_type]
+
+        if not isinstance(agent_cls, DeclarativeSpecProtocol):
+            raise AgentInitializationException(
+                f"Agent class '{agent_cls.__name__}' does not support declarative spec loading."
+            )
+
+        yaml_str = agent_cls.resolve_placeholders(yaml_str, settings, extras)
+        data = yaml.safe_load(yaml_str)
+
+        return await agent_cls.from_dict(
+            data,
+            kernel=kernel,
+            settings=settings,
+            **kwargs,
+        )
+
+    @staticmethod
+    async def create_from_dict(
+        data: dict,
+        *,
+        kernel: Kernel,
```

Comment:
all the agent constructors can do without a kernel and instead take a list of plugins, can we add that here and in the yaml one?

Reply:
Adding handling for one to pass in plugins without a kernel.